lib/fact_changes.rb

```diff
@@ -535,38 +522,31 @@ def _fact_operations(action_type, step, facts)
     operations = facts.map do |fact|
       modified_assets.push(fact.object_asset) if listening_to_predicate?(fact.predicate)
       Operation.new(:action_type => action_type, :step => step,
-        :asset=> fact.asset, :predicate => fact.predicate, :object => fact.object, object_asset: fact.object_asset)
+                    :asset => fact.asset, :predicate => fact.predicate, :object => fact.object, object_asset: fact.object_asset)
     end
 
     modified_assets.flatten.compact.uniq.each(&:touch)
     operations
   end
 
-  def all_values_are_new_records(hash)
-    hash.values.all? do |value|
-      (value.respond_to?(:new_record?) && value.new_record?)
-    end
+  def any_values_are_new_records(hash)
+    hash.values.any? { |value| value.try(:new_record?) }
   end
 
   def _instance_builder_for_import(klass, params_list, &block)
-
-    instances = params_list.map do |params_for_instance|
-      unless (params_for_instance.kind_of?(klass))
-        if (all_values_are_new_records(params_for_instance) ||
-          (!klass.exists?(params_for_instance)))
-          klass.new(params_for_instance)
-        end
-      else
-        if params_for_instance.new_record?
-          params_for_instance
-        end
+    instances = params_list.filter_map do |params_for_instance|
+      if params_for_instance.kind_of?(klass)
+        params_for_instance if params_for_instance.new_record?
+      elsif any_values_are_new_records(params_for_instance) || !klass.exists?(params_for_instance)
+        klass.new(params_for_instance)
       end
-    end.compact.uniq
+    end.uniq
     instances.each do |instance|
+      # Runs the callbacks (See https://github.com/zdennis/activerecord-import#callbacks)
       instance.run_callbacks(:save) { false }
       instance.run_callbacks(:create) { false }
     end
```

Comment:
I had hoped to be able to remove this on updating uuid generation, as that appeared to be the only before_save. However it appears that some other aspects of rails use these life-cyle hooks.